### PR TITLE
Implement backend for image analysis page.

### PR DESCRIPTION
**Pull request by Froggerson**

### CODE REVIEW

File: src/main/java/com/google/sps/servlets/ImageAnalysisServlet.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+ * When the user submits the form, Blobstore processes the file upload and then forwards the request
+ * to this servlet. This servlet can then analyze the image using the Vision API and return the labels
+ * for that image.
+ */
+@WebServlet("/image-analysis")
+public class ImageAnalysisServlet extends HttpServlet {
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
```
Comment (tylerhou) [inline src/main/java/com/google/sps/servlets/ImageAnalysisServlet.java:44]: No newline

File: src/main/java/com/google/sps/servlets/ImageAnalysisServlet.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get("image");
+
+    if (blobKeys == null || blobKeys.isEmpty()) {
+      return null;
+    }
+    BlobKey blobKey = blobKeys.get(0);
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    if (blobInfo.getSize() == 0) {
```
Comment (tylerhou) [inline src/main/java/com/google/sps/servlets/ImageAnalysisServlet.java:79]: Comment why we need this check and delete?

File: src/main/java/com/google/sps/servlets/ImageAnalysisServlet.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+   * BlobKey parameter.
+   */
+  private byte[] getBlobBytes(BlobKey blobKey) throws IOException {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    ByteArrayOutputStream outputBytes = new ByteArrayOutputStream();
+
+    int fetchSize = BlobstoreService.MAX_BLOB_FETCH_SIZE;
+    long currentByteIndex = 0;
+    boolean continueReading = true;
+    while (continueReading) {
```
Comment (tylerhou) [inline src/main/java/com/google/sps/servlets/ImageAnalysisServlet.java:97]: You can do while (true) and break instead, which avoids the continueReading variable.

File: src/main/java/com/google/sps/servlets/ImageAnalysisServlet.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+      // end index is inclusive, so we have to subtract 1 to get fetchSize bytes
+      byte[] b =
+          blobstoreService.fetchData(blobKey, currentByteIndex, currentByteIndex + fetchSize - 1);
+      outputBytes.write(b);
+      if (b.length < fetchSize) {
+        continueReading = false;
+      }
```
Comment (tylerhou) [inline src/main/java/com/google/sps/servlets/ImageAnalysisServlet.java:104]: There might be a bug here if the blob size is exactly a multiple of fetchSize. It depends on how fetchData is implemented -- does it throw an error if currentByteIndex is larger than the file? Reading the docs it shouldn't, so there should be no bug. https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/blobstore/BlobstoreService.html#fetchData-com.google.appengine.api.blobstore.BlobKey-long-long-

File: src/main/java/com/google/sps/servlets/ImageAnalysisServlet.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    PrintWriter out = response.getWriter();
+    String message = request.getParameter("message");
+    BlobKey blobKey = getBlobKey(request, "image");
+
+    if (blobKey == null) {
+      return;
+    }
+    String imageUrl = getUploadedFileUrl(blobKey);
```
Comment (tylerhou) [inline src/main/java/com/google/sps/servlets/ImageAnalysisServlet.java:52]: Never used.

**Review by tylerhou — APPROVED** [review]

File: src/main/java/com/google/sps/servlets/ImageAnalysisServlet.java (hunk 5 above)
Comment (Froggerson) [reply src/main/java/com/google/sps/servlets/ImageAnalysisServlet.java:52]: ok!

**Pull request merged**